src/txn/txn_recover.c

```diff
@@ -974,8 +1004,12 @@ __wt_txn_recover(WT_SESSION_IMPL *session, const char *cfg[])
          */
         WT_ERR(session->iface.checkpoint(&session->iface, "force=1"));
 
-    /* Initialize the connection's base write generation after rollback to stable. */
-    WT_ERR(__wt_metadata_init_base_write_gen(session));
+    /*
+     * Rollback to stable may have updated inconsistent transaction ids. Update the connection base
+     * write generation based on these new ids that may have been reset.
+     */
+    if (rts_executed)
+        WT_ERR(__recovery_correct_write_gen(session));
 
     /*
      * Update the open dhandles write generations and base write generation with the connection's
```

Comment:
I was wondering if we need to update the write gens of open dhandles in all the cases or only when RTS is executed.

Reply:
Yes. It is because of the history store.  As part of btree open, we set the btree base_write_gen as checkpoint write gen instead of connection base write gen for the history store, even if the RTS doesn't happen we need to reset them otherwise the visibility can go wrong. As part of resetting at the end of recovery to all open handles, it will be sure that no txnid is left out.

Reply:
Yeah, now I get it. Thanks, @kommiharibabu.